src/mlpack/methods/ann/layer/elu.hpp

```diff
@@ -205,19 +205,19 @@ class ELU
   /**
    * Computes the first derivative of the activation function.
    *
-   * @param x Input data.
+   * @param y Propagated data f(x).
    * @return f'(x)
    */
   double Deriv(const double y)
   {
-    return (y > 0) ? lambda : lambda * (y + alpha);
+    return (y > 0) ? lambda : y + lambda * alpha;
```

Comment:
The issue here is that `Deriv()` gets called from within the `Backward()` function, where `y` aliases `f(x)`. Hence the implementation goes slightly wrong by getting evaluated as `(f(x) > 0) ? lambda : f(x) + lambda * alpha;` whereas it should ideally be evaluated as `(x > 0) ? lambda : f(x) + lambda * alpha;`. I'm not sure how we can pass the original input `x` through `Backward()`. Suggestions are welcome.

Reply:
Also I think the previous implementation got lucky as `lambda` is only about `1.05...` numerically, and hence did not give a failure.

Reply:
I think we can get x from f(x) using f^{-1}=ln(\frac{x}{\lambda\cdot \alpha}+1).

Reply:
That would be computationally expensive, especially when we're using a large matrix for storing the numbers and for each number, the inverse would have to be computed. For an `N x N` matrix, that means `O(N^2)` extra computations, practically anything greater than `N = 10^4` should be noticeably slower. I was thinking more along the lines of locally storing the latest input of the `Forward()` pass, since `Backward()` would ideally not be called without the `Forward()` routine. @rcurtin ?

Reply:
Another solution might be to check the ``deterministic`` parameter and compute the derivative right away if necessary, that way we could avoid the extra copy. The solution also works for recurrent networks.

Reply:
@zoq Do you mean calling and saving the derivative from within the `Forward()` routine?

Reply:
Right if we are in train mode, I think you did something similar for the dropout class.